### PR TITLE
Improve clipping time with np_compat.clip

### PR DESCRIPTION
**Pull request by Lasloruhberg**
This PR introduces a numpy specific alias to xp.clip to improve performance, adressing Issue #444.
With respect to the basic implementation I removed array copies and new creations where possible and revert to the intrinsic np.clip.

The benchmark also implies that dask clip is quite below its possible performance, as it remains well below the other programs.

As far as I can tell this rewrite is compliant with the API standard, but please advise if not.
I am happy to receive feedback. 

## Current on main (run on a laptop, no GPU):
Batch size: 1024
Image shape: (128, 128, 3)
Measurement runs: 8

Skipped backends: cupy (unavailable)

| backend | dtype | avg ms | std ms | images/s | std images/s |
| --- | --- | ---: | ---: | ---: | ---: |
| numpy-native  | float32 | 230.097 | 20.311 | 4450.3 | 357.8 |
| numpy-native  | uint8 | 60.985 | 8.060 | 16790.9 | 1877.6 |
| numpy         | float32 | 4765.966 | 298.603 | 214.9 | 13.6 |
| numpy         | uint8 | 3194.393 | 266.443 | 320.6 | 27.0 |
| dask          | float32 | 3279.046 | 513.540 | 312.3 | 49.4 |
| dask          | uint8 | 1657.283 | 168.447 | 617.9 | 67.7 |
| jax           | float32 | 197.261 | 71.699 | 5191.1 | 1549.5 |
| jax           | uint8 | 38.581 | 7.265 | 26541.4 | 4950.1 |
| pytorch       | float32 | 160.713 | 54.031 | 6371.6 | 1940.3 |
| pytorch       | uint8 | 16.838 | 3.103 | 60814.3 | 13022.6 |

## Now on this branch

Batch size: 1024
Image shape: (128, 128, 3)
Measurement runs: 15

| backend | dtype | avg ms | std ms | images/s | std images/s |
| --- | --- | ---: | ---: | ---: | ---: |
| numpy-native | float32 | 156.005 | 27.627 | 6563.9 | 1260.4 |
| numpy-native | uint8 | 34.172 | 5.294 | 29965.8 | 4482.5 |
| numpy | float32 | 157.614 | 30.625 | 6496.9 | 1226.8 |
| numpy | uint8 | 41.640 | 7.193 | 24592.0 | 4832.1 |
| dask | float32 | 1675.356 | 204.533 | 611.2 | 81.3 |
| dask | uint8 | 1544.973 | 493.518 | 662.8 | 169.1 |
| jax | float32 | 81.755 | 30.772 | 12525.2 | 3866.9 |
| jax | uint8 | 42.276 | 66.549 | 24221.9 | 13959.8 |
| pytorch | float32 | 86.618 | 34.658 | 11822.1 | 3924.1 |
| pytorch | uint8 | 16.298 | 3.716 | 62829.0 | 12721.9 |

As evident the new clip version for numpy restores its native performance, while ensuring the same broadcasting and dtype preservation behaviour

Benchmark:
[benchmark.py](https://github.com/user-attachments/files/29627909/benchmark.py)

### CODE REVIEW

File: src/array_api_compat/numpy/_aliases.py [hunk 1]
```diff
@@ -106,6 +105,80 @@ def astype(
...
 
 
+def clip(
+    x: Array,
+    /,
+    min: float | Array | None = None,
+    max: float | Array | None = None,
+    out: Array | None = None,
+    **kwargs,
+) -> Array:
```
[the remark this answers is not in the document]
Comment (Lasloruhberg) [reply src/array_api_compat/numpy/_aliases.py:114]: Removed out, if in kwargs pop it from there and use internally in function only

File: src/array_api_compat/numpy/_aliases.py [hunk 2]
```diff
@@ -115,7 +196,9 @@ def count_nonzero(
 ) -> Array:
     # NOTE: this is currently incorrectly typed in numpy, but will be fixed in
     # numpy 2.2.5 and 2.3.0: https://github.com/numpy/numpy/pull/28750
-    result = cast("Any", np.count_nonzero(x, axis=axis, keepdims=keepdims))  # pyright: ignore[reportArgumentType, reportCallIssue]
+    result = cast(
```
Comment (ev-br) [inline src/array_api_compat/numpy/_aliases.py]: please roll back unrelated changes

File: tests/test_numpy.py [hunk 3]
```diff
@@ -10,6 +10,84 @@
 
 from array_api_compat import is_array_api_obj
 
+
+def test_numpy_clip_out_and_broadcast():
+    from array_api_compat import numpy as xp
```
Comment (ev-br) [inline tests/test_numpy.py]: AI loves inline imports, but there's really no need for that here. Could you please move them to the top of the file

**Review by ev-br — COMMENTED** [review]
Thanks.

Given the (somewhat cryptic) comments in the "generic" clip,
https://github.com/data-apis/array-api-compat/blob/main/src/array_api_compat/common/_aliases.py#L395, could you please add some tests to probe the uint64->float64 promotion that this comments hint at?

**Comment by Lasloruhberg** [comment]
I have reworked the test.
In this code the input array always keeps its dtype.
Only the min/max bounds are cast to the target dtype, which is now rigourously protected against overflows in the case of an Integer array as an input
The tests ensure:
- Integer Array clip: Downcasting integer dtypes, which are now rigourous for any bounds input (Small caveat is that it relies on numpys internal conversion of float to integer, which might lead to unexpected behaviour)
- Float array clip: Downcasting float dtypes, numpy replaces them with +- inf, which then works as expected

**Comment by Lasloruhberg** [comment]
Unsure why tests fail, seems to be a server issue:
```
  WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLError(1, '[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:1017)'))': /whl/cpu/torch-2.12.1%2Bcpu-cp310-cp310-manylinux_2_28_x86_64.whl.metadata
ERROR: Could not install packages due to an OSError: HTTPSConnectionPool(host='download-r2.pytorch.org', port=443): Max retries exceeded with url: /whl/cpu/torch-2.12.1%2Bcpu-cp310-cp310-manylinux_2_28_x86_64.whl.metadata (Caused by SSLError(SSLError(1, '[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:1017)')))

```



Would suggest to wait till tomorrow to see if it fixes itself and then review this

**Comment by Lasloruhberg** [comment]
Now works stable for all of the inputs, such as lists, tuple, arrays
For the bounds:
- Downcasting to integers is safe, by ensuring proper clipping in case of out of bound floats/ints,
- Downcasting to lower bit floats is naturally handled by numpys asarray with a fixed dtype 
    (replaces out of bounds with +- inf)

The array that should be clipped is never changed in its dtype

**Review by ev-br — APPROVED** [review]
`clip` is a surprisingly tricky function, with weird edge cases, so I won't be surprised if usage discovers some of them. Hard to tell without actually using it, so let's give this a go. I ran locally the clip test of `array-api-tests` with a large number of examples, and also scipy tests against this branch. No problems so far, so let's roll with it. Thank you @Lasloruhberg
